File: src/app/services/exchange-rate.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

export interface ExchangeRateData {
  date: Date;
  currency: string;
  buyingRate: number;
  sellingRate: number;
  source?: string;
  averageRate: number;
}

@Injectable({
  providedIn: 'root'
})
export class ExchangeRateService {
  private readonly baseUrl = 'https://www.cbsl.gov.lk/en/rates-and-indicators/exchange-rates/daily-buy-and-sell-exchange-rates';

  constructor(private http: HttpClient) {}

  /**
   * Fetches exchange rate for a specific currency and date range
   */
  getExchangeRateForDateRange(startDate: Date, endDate: Date, currency: string): Observable<ExchangeRateData> {
    return new Observable(observer => {
      this.getRealCBSLData(startDate, endDate, currency)
        .then(data => {
          if (data) {
            observer.next(data);
            observer.complete();
          } else {
            const dateStr = startDate.getTime() === endDate.getTime() 
              ? this.formatDate(startDate)
              : `${this.formatDate(startDate)} to ${this.formatDate(endDate)}`;
            observer.error(new Error(`No exchange rate available for ${dateStr} from Central Bank of Sri Lanka (CBSL). Exchange rates are only available for working days when the bank publishes them. Please try a different date or check if the selected date is a working day.`));
          }
        })
        .catch(error => {
          console.error('Error fetching exchange rate:', error);
          const dateStr = startDate.getTime() === endDate.getTime() 
            ? this.formatDate(startDate)
            : `${this.formatDate(startDate)} to ${this.formatDate(endDate)}`;
          observer.error(new Error(`No exchange rate available for ${dateStr} from Central Bank of Sri Lanka (CBSL). Please try a different date or check if the selected date is a working day.`));
        });
    });
  }

  /**
   * Fetches exchange rate for a specific currency and date
   */
  getExchangeRate(date: Date, currency: string): Observable<ExchangeRateData> {
    return this.getExchangeRateForDateRange(date, date, currency);
  }

  private async getRealCBSLData(startDate: Date, endDate: Date, currency: string): Promise<ExchangeRateData | null> {
    try {
      const startDateStr = this.formatDate(startDate);
      const endDateStr = this.formatDate(endDate);
      
      // Store the original requested date for exact matching
      const originalRequestedDate = startDate.getTime() === endDate.getTime() ? startDate : endDate;
      
      // If looking for a specific date, expand the search range to find data, but we'll filter for exact date
      let actualStartDateStr = startDateStr;
      let actualEndDateStr = endDateStr;
      
      if (startDate.getTime() === endDate.getTime()) {
        const expandedStartDate = new Date(startDate);
        expandedStartDate.setDate(startDate.getDate() - 3);
        
        const expandedEndDate = new Date(endDate);
        expandedEndDate.setDate(endDate.getDate() + 3);
        
        actualStartDateStr = this.formatDate(expandedStartDate);
        actualEndDateStr = this.formatDate(expandedEndDate);
      }

      // Use our proxy server to fetch CBSL data
      const requestBody = {
        startDate: actualStartDateStr,
        endDate: actualEndDateStr,
        currency: currency,
        exactDate: this.formatDate(originalRequestedDate) // Send the exact date we want
      };

      const response = await this.http.post<any>('http://localhost:3001/api/cbsl/exchange-rate', requestBody).toPromise();

      if (response && response.success) {
        console.log('Successfully received CBSL data via proxy:', response.data);
        return {
          date: new Date(response.data.date),
          currency: response.data.currency,
          buyingRate: response.data.buyingRate,
          sellingRate: response.data.sellingRate,
          source: response.data.source,
          averageRate: response.data.averageRate
        };
      } else {
        console.log('CBSL proxy returned no data:', response?.error);
        return null;
      }
    } catch (error) {
      console.error('Error fetching CBSL data via proxy:', error);
      return null;
    }
  }

  private parseCBSLResponse(responseBody: string, startDate: Date, endDate: Date, currency: string): ExchangeRateData | null {
    try {
      // Create a temporary DOM element to parse the HTML
      const parser = new DOMParser();
      const document = parser.parseFromString(responseBody, 'text/html');
      const tables = document.querySelectorAll('table');
      
      if (tables.length === 0) {
        console.log('No tables found in CBSL response');
        return null;
      }
      
      const targetDate = startDate.getTime() === endDate.getTime() ? startDate : endDate;
      
      // Look for the exchange rate table
      for (const table of tables) {
        const rows = table.querySelectorAll('tr');
        if (rows.length < 2) continue;
        
        // Check if this looks like an exchange rate table
        const headerRow = rows[0];
        const headerCells = headerRow.querySelectorAll('td, th');
        const headerTexts = Array.from(headerCells).map(cell => cell.textContent?.trim().toLowerCase() || '');
        
        // Look for table with Date, Buy Rate, Sell Rate columns
        const hasDate = headerTexts.some(text => text.includes('date'));
        const hasBuy = headerTexts.some(text => text.includes('buy'));
        const hasSell = headerTexts.some(text => text.includes('sell'));
        
        if (hasDate && hasBuy && hasSell) {
          const rates: any[] = [];
          
          for (let i = 1; i < rows.length; i++) {
            const row = rows[i];
            const cells = row.querySelectorAll('td, th');
            const cellTexts = Array.from(cells).map(cell => cell.textContent?.trim() || '');
            
            if (cellTexts.length >= 3) {
              try {
                const dateStr = cellTexts[0];
                const buyRateStr = cellTexts[1].replace(/,/g, '');
                const sellRateStr = cellTexts[2].replace(/,/g, '');
                
                const date = new Date(dateStr);
                const buyRate = parseFloat(buyRateStr);
                const sellRate = parseFloat(sellRateStr);
                
                if (!isNaN(date.getTime()) && !isNaN(buyRate) && !isNaN(sellRate)) {
                  rates.push({
                    date: date,
                    buyingRate: buyRate,
                    sellingRate: sellRate,
                    daysDifference: Math.abs(Math.floor((date.getTime() - targetDate.getTime()) / (1000 * 60 * 60 * 24)))
                  });
                }
              } catch (e) {
                console.error(`Error parsing row ${i}:`, e);
              }
            }
          }
          
          if (rates.length > 0) {
            // Find exact date match
            const exactMatch = rates.filter(rate => {
              const rateDate = rate.date;
              return rateDate.getFullYear() === targetDate.getFullYear() && 
                     rateDate.getMonth() === targetDate.getMonth() && 
                     rateDate.getDate() === targetDate.getDate();
            });
            
            if (exactMatch.length > 0) {
              const exactRate = exactMatch[0];
              console.log(`Found exact date match for ${this.formatDate(targetDate)}`);
              
              return {
                date: exactRate.date,
                currency: currency,
                buyingRate: exactRate.buyingRate,
                sellingRate: exactRate.sellingRate,
                source: 'Central Bank of Sri Lanka (CBSL)',
                averageRate: (exactRate.buyingRate + exactRate.sellingRate) / 2
              };
            }
            
            console.log(`No exact exchange rate found for ${this.formatDate(targetDate)}`);
            return null;
          }
        }
      }
      
      console.log('No valid exchange rate table found in CBSL response');
      return null;
      
    } catch (error) {
      console.error('Error parsing CBSL response:', error);
      return null;
    }
  }

  private formatDate(date: Date): string {
    const year = date.getFullYear();
    const month = (date.getMonth() + 1).toString().padStart(2, '0');
    const day = date.getDate().toString().padStart(2, '0');
    return `${year}-${month}-${day}`;
  }

  private getCurrencyMapping(currency: string): string {
    const mappings: { [key: string]: string } = {
      'USD': 'USD~US Dollar',
      'SGD': 'SGD~Singapore Dollar',
      'EUR': 'EUR~Euro',
      'GBP': 'GBP~Pound Sterling',
      'JPY': 'JPY~Japanese Yen',
      'AUD': 'AUD~Australian Dollar',
      'CAD': 'CAD~Canadian Dollar',
      'CHF': 'CHF~Swiss Franc',
      'CNY': 'CNY~Chinese Yuan',
      'INR': 'INR~Indian Rupee'
    };
    return mappings[currency] || `${currency}~${currency}`;
  }

  /**
   * Get list of supported currencies
   */
  getSupportedCurrencies(): { code: string, name: string }[] {
    return [
      { code: 'USD', name: 'United States Dollar' },
      { code: 'EUR', name: 'Euro' },
      { code: 'GBP', name: 'British Pound' },
      { code: 'AUD', name: 'Australian Dollar' },
      { code: 'CAD', name: 'Canadian Dollar' },
      { code: 'CHF', name: 'Swiss Franc' },
      { code: 'CNY', name: 'Renminbi' },
      { code: 'JPY', name: 'Yen' },
      { code: 'SGD', name: 'Singapore Dollar' }
    ];
  }

  /**
   * Returns a fallback rate when live data is not available (for testing/development)
   */
  private getFallbackRate(date: Date, currency: string): ExchangeRateData {
    console.log(`Using fallback exchange rate for ${currency} on ${this.formatDate(date)}`);
    
    // Base rates (approximate 2024-2025 rates)
    const baseRates: { [key: string]: number } = {
      'USD': 300.0,
      'SGD': 220.0,
      'EUR': 330.0,
      'GBP': 380.0,
      'JPY': 2.1,
      'AUD': 195.0,
      'CAD': 220.0,
      'CHF': 335.0,
      'CNY': 42.0,
      'INR': 3.6
    };
    
    // Simulate slight variation based on date for realism
    const dayOfMonth = date.getDate();
    const baseRate = baseRates[currency] || 250.0;
    const variation = (dayOfMonth % 5) * 0.5 - 1.0; // Small daily variation
    
    const buyingRate = baseRate + variation - 1.5;
    const sellingRate = baseRate + variation + 1.5;
    
    return {
      date: date,
      currency: currency,
      buyingRate: parseFloat(buyingRate.toFixed(2)),
      sellingRate: parseFloat(sellingRate.toFixed(2)),
      source: 'Fallback Rate (CBSL data temporarily unavailable)',
      averageRate: parseFloat(((buyingRate + sellingRate) / 2).toFixed(2))
    };
  }
}

File: reference/exchange_rate_service.dart
import 'package:http/http.dart' as http;
import 'package:html/parser.dart' as parser;

class ExchangeRateService {
  static const String baseUrl = 'https://www.cbsl.gov.lk/en/rates-and-indicators/exchange-rates/daily-buy-and-sell-exchange-rates';
  
  /// Fetches SGD to LKR exchange rate for a specific date range
  Future<ExchangeRateData> getExchangeRateForDateRange(DateTime startDate, DateTime endDate) async {
    try {
      // Try to get real CBSL data first
      final realData = await _getRealCBSLData(startDate, endDate);
      if (realData != null) {
        return realData;
      }
      
      // Throw an error if we can't get actual exchange rate data for the exact date
      final dateStr = startDate.isAtSameMomentAs(endDate) 
          ? startDate.toIso8601String().split('T')[0] 
          : '${startDate.toIso8601String().split('T')[0]} to ${endDate.toIso8601String().split('T')[0]}';
      throw Exception('No exchange rate available for $dateStr from Central Bank of Sri Lanka (CBSL). Exchange rates are only available for working days when the bank publishes them. Please try a different date or check if the selected date is a working day.');
    } catch (e) {
      print('Error fetching exchange rate: $e');
      // Re-throw the error to be handled by the calling code
      rethrow;
    }
  }

  Future<ExchangeRateData?> _getRealCBSLData(DateTime startDate, DateTime endDate) async {
    try {
      // Format dates as YYYY-MM-DD for CBSL form
      final startDateStr = '${startDate.year}-${startDate.month.toString().padLeft(2, '0')}-${startDate.day.toString().padLeft(2, '0')}';
      final endDateStr = '${endDate.year}-${endDate.month.toString().padLeft(2, '0')}-${endDate.day.toString().padLeft(2, '0')}';
      
      // If looking for a specific date (start and end are the same), 
      // expand the search range to find the closest available rate
      String actualStartDateStr = startDateStr;
      String actualEndDateStr = endDateStr;
      
      if (startDate.isAtSameMomentAs(endDate)) {
        // Look for data within a 7-day range around the target date
        final expandedStartDate = startDate.subtract(const Duration(days: 3));
        final expandedEndDate = endDate.add(const Duration(days: 3));
        
        actualStartDateStr = '${expandedStartDate.year}-${expandedStartDate.month.toString().padLeft(2, '0')}-${expandedStartDate.day.toString().padLeft(2, '0')}';
        actualEndDateStr = '${expandedEndDate.year}-${expandedEndDate.month.toString().padLeft(2, '0')}-${expandedEndDate.day.toString().padLeft(2, '0')}';
      }
      
      // Create the exact form data structure from CBSL iframe
      final formData = {
        'lookupPage': 'lookup_daily_exchange_rates.php',
        'startRange': '2006-11-11',
        'rangeType': 'dates',
        'txtStart': actualStartDateStr,
        'txtEnd': actualEndDateStr,
        'chk_cur[]': 'SGD~Singapore Dollar',
        'submit_button': 'Submit',
      };
      
      // Submit to CBSL iframe endpoint
      final response = await http.post(
        Uri.parse('https://www.cbsl.gov.lk/cbsl_custom/exratestt/exrates_resultstt.php'),
        headers: {
          'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
          'Content-Type': 'application/x-www-form-urlencoded',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
          'Accept-Language': 'en-US,en;q=0.5',
          'Connection': 'keep-alive',
          'Referer': 'https://www.cbsl.gov.lk/cbsl_custom/exratestt/exratestt.php',
          'Origin': 'https://www.cbsl.gov.lk',
        },
        body: formData.entries.map((e) => '${Uri.encodeComponent(e.key)}=${Uri.encodeComponent(e.value)}').join('&'),
      );
      
      if (response.statusCode == 200) {
        return _parseCBSLResponse(response.body, startDate, endDate);
      } else {
        print('CBSL request failed with status: ${response.statusCode}');
        return null;
      }
      
    } catch (e) {
      print('Error fetching CBSL data: $e');
      return null;
    }
  }

  ExchangeRateData? _parseCBSLResponse(String responseBody, DateTime startDate, DateTime endDate) {
    try {
      final document = parser.parse(responseBody);
      final tables = document.querySelectorAll('table');
      
      if (tables.isEmpty) {
        print('No tables found in CBSL response');
        return null;
      }
      
      // The target date (what we're actually looking for)
      final targetDate = startDate.isAtSameMomentAs(endDate) ? startDate : endDate;
      
      // Look for the exchange rate table
      for (final table in tables) {
        final rows = table.querySelectorAll('tr');
        if (rows.length < 2) continue;
        
        // Check if this looks like an exchange rate table
        final headerRow = rows[0];
        final headerCells = headerRow.querySelectorAll('td, th');
        final headerTexts = headerCells.map((cell) => cell.text.trim().toLowerCase()).toList();
        
        // Look for table with Date, Buy Rate, Sell Rate columns
        final hasDate = headerTexts.any((text) => text.contains('date'));
        final hasBuy = headerTexts.any((text) => text.contains('buy'));
        final hasSell = headerTexts.any((text) => text.contains('sell'));
        
        if (hasDate && hasBuy && hasSell) {
          // Parse exchange rate data from this table
          final rates = <Map<String, dynamic>>[];
          
          for (int i = 1; i < rows.length; i++) {
            final row = rows[i];
            final cells = row.querySelectorAll('td, th');
            final cellTexts = cells.map((cell) => cell.text.trim()).toList();
            
            if (cellTexts.length >= 3) {
              try {
                final dateStr = cellTexts[0];
                final buyRateStr = cellTexts[1].replaceAll(',', '');
                final sellRateStr = cellTexts[2].replaceAll(',', '');
                
                final date = DateTime.tryParse(dateStr);
                final buyRate = double.tryParse(buyRateStr);
                final sellRate = double.tryParse(sellRateStr);
                
                if (date != null && buyRate != null && sellRate != null) {
                  rates.add({
                    'date': date,
                    'buyingRate': buyRate,
                    'sellingRate': sellRate,
                    'daysDifference': (date.difference(targetDate).inDays).abs(),
                  });
                }
              } catch (e) {
                print('Error parsing row $i: $e');
              }
            }
          }
          
          if (rates.isNotEmpty) {
            // Only try to find exact date match - no approximations
            final exactMatch = rates.where((rate) {
              final rateDate = rate['date'] as DateTime;
              return rateDate.year == targetDate.year && 
                     rateDate.month == targetDate.month && 
                     rateDate.day == targetDate.day;
            }).toList();
            
            if (exactMatch.isNotEmpty) {
              final exactRate = exactMatch.first;
              print('Found exact date match for ${targetDate.toIso8601String().split('T')[0]}');
              
              return ExchangeRateData(
                date: exactRate['date'] as DateTime,
                currency: 'SGD',
                buyingRate: exactRate['buyingRate'] as double,
                sellingRate: exactRate['sellingRate'] as double,
                source: 'Central Bank of Sri Lanka (CBSL)',
              );
            }
            
            // No exact match found - return null to indicate failure
            print('No exact exchange rate found for ${targetDate.toIso8601String().split('T')[0]}');
            return null;
          }
        }
      }
      
      print('No valid exchange rate table found in CBSL response');
      return null;
      
    } catch (e) {
      print('Error parsing CBSL response: $e');
      return null;
    }
  }
  
  /// Fetches SGD to LKR exchange rate for a specific date (backward compatibility)
  Future<ExchangeRateData> getExchangeRate(DateTime date) async {
    // For single date requests, use a 3-day range ending on the requested date
    final fromDate = date.subtract(Duration(days: 2));
    return getExchangeRateForDateRange(fromDate, date);
  }
  
  /// Returns a fallback rate when live data is not available (for testing/development only)
  ExchangeRateData _getFallbackRate(DateTime date) {
    // Use more realistic current rates as fallback
    // These rates are based on approximate 2024-2025 rates
    print('Using fallback exchange rate for ${date.toIso8601String().split('T')[0]}');
    
    // Simulate slight variation based on date for realism
    final dayOfMonth = date.day;
    final baseRate = 243.0;
    final variation = (dayOfMonth % 5) * 0.5 - 1.0; // Small daily variation
    
    final buyingRate = baseRate + variation - 1.5;
    final sellingRate = baseRate + variation + 1.5;
    
    return ExchangeRateData(
      date: date,
      currency: 'SGD',
      buyingRate: double.parse(buyingRate.toStringAsFixed(2)),
      sellingRate: double.parse(sellingRate.toStringAsFixed(2)),
      source: 'Estimated Rate (CBSL data unavailable)',
    );
  }
  
  /// Get fallback rate with error indication (for emergency use only)
  Future<ExchangeRateData> getFallbackRateWithWarning(DateTime date) async {
    print('Using emergency fallback exchange rate');
    return _getFallbackRate(date);
  }
  
  /// Alternative method: Get approximate exchange rate using current market data
  Future<ExchangeRateData?> getApproximateRate(DateTime date) async {
    print('Getting approximate exchange rate');
    return _getFallbackRate(date);
  }
}

class ExchangeRateData {
  final DateTime date;
  final String currency;
  final double buyingRate;
  final double sellingRate;
  final String? source;
  
  ExchangeRateData({
    required this.date,
    required this.currency,
    required this.buyingRate,
    required this.sellingRate,
    this.source,
  });
  
  double get averageRate => (buyingRate + sellingRate) / 2;
  
  Map<String, dynamic> toJson() {
    return {
      'date': date.toIso8601String(),
      'currency': currency,
      'buyingRate': buyingRate,
      'sellingRate': sellingRate,
      'source': source,
    };
  }
  
  factory ExchangeRateData.fromJson(Map<String, dynamic> json) {
    return ExchangeRateData(
      date: DateTime.parse(json['date']),
      currency: json['currency'],
      buyingRate: json['buyingRate'].toDouble(),
      sellingRate: json['sellingRate'].toDouble(),
      source: json['source'],
    );
  }
  
  @override
  String toString() {
    return 'ExchangeRateData(date: $date, currency: $currency, buying: $buyingRate, selling: $sellingRate, source: $source)';
  }
}
